google/auth/impersonated_credentials.py

```diff
@@ -205,7 +205,13 @@ def __init__(
         super(Credentials, self).__init__()
 
         self._source_credentials = copy.copy(source_credentials)
-        self._source_credentials._scopes = _IAM_SCOPE
+        # Service account source credentials must have the _IAM_SCOPE
+        # added to refresh correctly. User credentials cannot have
+        # their original scopes modified.
+        try:
+            self._source_credentials = self._source_credentials.with_scopes(_IAM_SCOPE)
```

Comment:
I think it'd be better to check the `Credentials` type and add the `scope` depending on that, rather than catching an error. This makes the logic more obvious and less error prone.

Reply:
Only the `Credentials` which inherits `google.auth.Credentials.Scoped` has the `with_scopes` method (service account credentials class inherits Scoped, and user credentials class is not). So you can use the following code instead of the try/except block:

```
if isinstance(self._source_credentials, credentials.Scoped):
    self._source_credentials = self._source_credentials.with_scopes(_IAM_SCOPE)
```

Reply:
Modified, thank you for the suggestion @arithmetic1728!